trio/_core/_gen_exports.py

```diff
@@ -0,0 +1,147 @@
+    """
+    doc = func.body[0]
+    if isinstance(doc, ast.Expr):
+        if hasattr(doc, 'value'):
+            if hasattr(doc.value, 's'):
+                return doc.value.s
+    return ""
+
+
+def gen_exports():
+    """ Sync the existing functions with the list
+    of exported functions imported from _public
+    """
+
+    imports = """from ._run import GLOBAL_RUN_CONTEXT, Runner
+from ._ki import LOCALS_KEY_KI_PROTECTION_ENABLED
+
+
+"""
+
+    for module in get_export_modules_by_dir(SOURCE_TREE):
+        full_path = os.path.join(
+            module.module_path, '_public' + module.module_file
+        )
+        try:
+            os.remove(full_path)
+        except FileNotFoundError:
+            pass
+        with open(full_path, 'w') as pub_file:
+            pub_file.writelines(imports)
+
+    # Get all modules we have classes with methods to export in the directory path
+    trees = get_export_modules_by_dir(SOURCE_TREE)
+
+    # Get all methods we want to export
+    methods = [meth for tree in trees for meth in get_public_methods(tree)]
+
+    # for m in methods:
+    #     try:
+    #         os.remove(os.path.join(m.module_path, '_public' + m.module_file))
+    #     except FileNotFoundError:
+    #         pass
+
+    # Get the module and the class of the method as well as the defaults
+    for m in methods:
+        if m.parent.name == 'Runner':
+            ctx = 'runner'
+        else:
+            ctx = "runner.io_manager"
+        # Remove self argument
+        m.args.args.pop(0)
+        # Remove decorators
+        m.decorator_list = []
+        # Replace _NO_SEND)
+        for default in m.args.defaults:
+            if isinstance(default, ast.Name):
+                default.id = 'Runner._NO_SEND'
+        args = astor.to_source(m.args)
+        arg_list = args.split(',')
+        new_arg_list = []
+        for arg in arg_list:
+            if '=' in arg:
+                key, value = arg.split('=')
+                key = key.strip()
+                new_arg_list.append('='.join([key, key]))
+            else:
+                new_arg_list.append(arg.strip())
+        new_args = ', '.join(new_arg_list)
+        # Strip PUBLIC keyword from docs
+        doc_string = m.body[0].value.s.replace('PUBLIC ', '')
+        m.body[0].value.s = doc_string
+        # Remove function body
+        del m.body[1:]
+        # Create new function body
+        template = """locals()[LOCALS_KEY_KI_PROTECTION_ENABLED] = True
+try:
+    return GLOBAL_RUN_CONTEXT.{}.{}
+except AttributeError:
+    raise RuntimeError('must be called from async context')
+    """.format(ctx, m.name + '(' + new_args + ')')
+        # Assemble function
+        ast_method = ast.parse(template)
+        m.body.extend(ast_method.body)
+        source = astor.to_source(m).replace('async ', '') + '\n\n'
+        with open(
+            os.path.join(m.module_path, '_public' + m.module_file), 'a'
+        ) as pub_file:
+            pub_file.writelines(source)
```

Comment:
We're also going to want a version of this that generates the text and then checks whether it matches what's already on disk (with just a simple string comparison). We'll run this in CI.

Reply:
Oh, I had a different way in mind. The append is only because I do the imports first and then append to the file which has just the imports. I do not bother what was on disk before, I just create a new version every time. Why check the old version?

Reply:
@njsmith I now have the generated file formatted via yapf. Before I write the comparison, where should the current `_gen_exports.py` file live? What should be its name?
I still have to write the test though, but I'd like to know the final destination first, as it might not be part of code coverage.